docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+in https://github.com/conduitio/conduit-plugin. This repository is the only common thing between Conduit and a plugin,
+meaning that Conduit uses the interface to interact with plugins and plugins implement the interface.
+
+The proto files define the messages and gRPC interface that needs to be implemented by the connector plugin. The
+interface is uploaded to the [buf schema registry](https://buf.build/conduitio/conduit-plugin), which allows us to
+easily generate the code in any programming language. For Go we
+use [remote generation](https://docs.buf.build/bsr/remote-generation/overview) so we don’t have to manually generate any
+code.
+
+Alongside the proto files, the repository contains helper structs and methods that hide the gRPC implementation from the
+user without changing any semantics. No other code needs to know that the communication happens through gRPC (i.e.
+neither Conduit nor the plugin SDK).
+
+There are 3 gRPC services that a plugin needs to implement - `SourcePlugin`, `DestinationPlugin` and `SpecifierPlugin`.
+Notice that request/response messages are defined in a very verbose way to give us the ability to adapt them in the
+future as we see fit without creating a whole new plugin interface.
+
+The source plugin interface looks as follows:
+
+```protobuf
+service SourcePlugin {
+  rpc Configure(Source.Configure.Request) returns (Source.Configure.Response);
+  rpc Start(Source.Start.Request) returns (Source.Start.Response);
+  rpc Run(stream Source.Run.Request) returns (stream Source.Run.Response);
+  rpc Stop(Source.Stop.Request) returns (Source.Stop.Response);
+  rpc Teardown(Source.Teardown.Request) returns (Source.Teardown.Response);
+}
+```
+
+Functions will be called in the order in which they are defined.
+
+- `Configure` - the plugin needs to validate the configuration it receives and either store the configuration and return
+  no error or discard it and return an error explaining why the config is invalid. This function serves two purposes:
+    - Config validation - Conduit calls `Configure` when the connector is first created or when the configuration is
+      updated to validate the configuration. In this case the next call is `Teardown` and the plugin is stopped.
+    - Configuring the plugin - Conduit calls `Configure` when the pipeline is started to provide the plugin with its
+      config. The next call after a successful response is `Start`.
+- `Start` - with a call to this function Conduit signals to the plugin that it wants it to start running. The request
+  will contain the position at which the plugin should start running (the position might be empty in case the pipeline
+  is started for the first time). The plugin is expected to open any connections needed to fetch records. In case of a
+  successful response, the next call will be `Run`.
+- `Run` - the important thing to notice here is that this function opens a bidirectional stream between Conduit and the
```

Comment:
> How about about we unify Start() and Run()?

We can't do that in a nice way in the gRPC interface, but we actually do that in the [internal plugin interface](https://github.com/ConduitIO/conduit/pull/109/files#diff-d4abb475517ba51087ef20b0522374775fb256dffe06ab36a81a2e018f1767f7R127-R143) in Conduit (notice there is only `Start`). The thing is that `Start` is a simple request/response endpoint, which allows us to send initialization data to the plugin (e.g. the position where to start from), while `Run` opens a bidirectional stream, where the requests are acks and responses are records. In other words - the first message we want to send to the plugin is not an "ack", so we need a separate endpoint to deliver it.

Now, what we _could_ do is send the position through `Configure`, and I considered it, but ultimately decided to rather create a separate endpoint, because `Configure` is also used to validate the configuration a user supplies to the plugin, and the position doesn't play a role there. Another benefit of having it separate is that we expect the plugin will open the connection to any 3rd party resources in `Start` so if that call succeeds we can safely open the stream. I think it makes it just a bit easier to spot an initialization error, because it's a clean and separate step.